ovs-p4rt/ovs_p4rt.cc

```diff
@@ -113,107 +113,73 @@ void PrepareFdbTxVlanTableEntry(p4::v1::TableEntry* table_entry,
 
-  match1->mutable_exact()->set_value(EncodeByteValue(1, 0));
+  match1->mutable_exact()->set_value(EncodeByteValue(1, learn_info.bridge_id));
 #endif
 
   if (insert_entry) {
     auto table_action = table_entry->mutable_action();
     auto action = table_action->mutable_action();
-    action->set_action_id(GetActionId(p4info, L2_FWD_TX_TABLE_ACTION_L2_FWD));
-    {
-      auto param = action->add_params();
-      param->set_param_id(GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_L2_FWD,
-                                     ACTION_L2_FWD_PARAM_PORT));
+    if (learn_info.vlan_info.port_vlan_mode == P4_PORT_VLAN_NATIVE_UNTAGGED) {
+      action->set_action_id(
+          GetActionId(p4info, L2_FWD_TX_TABLE_ACTION_REMOVE_VLAN_AND_FWD));
+      {
+        auto param = action->add_params();
+        param->set_param_id(
+            GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_REMOVE_VLAN_AND_FWD,
+                       ACTION_REMOVE_VLAN_AND_FWD_PARAM_PORT_ID));
+        auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
+        param->set_value(EncodeByteValue(1, port_id));
+      }
+      {
+        auto param = action->add_params();
+        param->set_param_id(
+            GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_REMOVE_VLAN_AND_FWD,
+                       ACTION_REMOVE_VLAN_AND_FWD_PARAM_VLAN_PTR));
+        param->set_value(EncodeByteValue(1, learn_info.vlan_info.port_vlan));
+      }
+    } else if (learn_info.vlan_info.port_vlan_mode ==
+               P4_PORT_VLAN_NATIVE_TAGGED) {
+      action->set_action_id(
+          GetActionId(p4info, L2_FWD_TX_TABLE_ACTION_ADD_VLAN_AND_FWD));
+      {
+        auto param = action->add_params();
+        param->set_param_id(GetParamId(p4info,
+                                       L2_FWD_TX_TABLE_ACTION_ADD_VLAN_AND_FWD,
+                                       ACTION_ADD_VLAN_AND_FWD_PARAM_PORT_ID));
+        auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
+        param->set_value(EncodeByteValue(1, port_id));
+      }
+      {
+        auto param = action->add_params();
+        param->set_param_id(GetParamId(p4info,
+                                       L2_FWD_TX_TABLE_ACTION_ADD_VLAN_AND_FWD,
+                                       ACTION_ADD_VLAN_AND_FWD_PARAM_VLAN_PTR));
+        param->set_value(EncodeByteValue(1, learn_info.vlan_info.port_vlan));
+      }
+    } else {
+      action->set_action_id(GetActionId(p4info, L2_FWD_TX_TABLE_ACTION_L2_FWD));
+      {
+        auto param = action->add_params();
+        param->set_param_id(GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_L2_FWD,
+                                       ACTION_L2_FWD_PARAM_PORT));
 #if defined(DPDK_TARGET)
-      auto port_id = learn_info.vln_info.vlan_id - 1;
+        auto port_id = learn_info.vln_info.vlan_id - 1;
 #elif defined(ES2K_TARGET)
-      auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
+        auto port_id = learn_info.mac_addr[1] + ES2K_VPORT_ID_OFFSET;
```

Comment:
Here port_id is forwarding vsi_id and we are using a workaround to get the vsi id from 2nd octet of mac address + offest (16). Please add a comment explaining this and also a TODO to revisit this if IMC logic for VSI is changes

Reply:
I can stiill add a comment regarding port ID